File: cprdspy/CPR_matplotlib/Circles/circle.py
import numpy as np
import matplotlib.pyplot as plt


def circle(
    radius=1,
    color="#0F0",
    alpha=1,
    center=(0, 0),
    points=100,
    linestyle="-",
    linewidth=1,
    label=None,
    marker=None,
    markersize=5,
    markerfacecolor="r",
    markeredgecolor="k",
    markeredgewidth=1,
    ax=None,
    **kwargs,
):
    """
    绘制一个圆。

    参数:
        radius (float): 圆的半径，默认为1。
        color (str): 圆的颜色，默认为绿色("#0F0")。
        alpha (float): 透明度，默认为1（不透明）。
        center (tuple): 圆心坐标，默认为(0, 0)。
        points (int): 用于绘制圆的点数，默认为100。
        linestyle (str): 线型，默认为实线("-")。
        linewidth (int): 线宽，默认为1。
        label (str): 图例标签，默认为None。
        marker (str): 标记样式，默认为None。
        markersize (int): 标记大小，默认为5。
        markerfacecolor (str): 标记填充颜色，默认为红色("r")。
        markeredgecolor (str): 标记边缘颜色，默认为黑色("k")。
        markeredgewidth (int): 标记边缘宽度，默认为1。
        ax (matplotlib.axes.Axes): 目标坐标轴，默认为None（使用当前坐标轴）。
        **kwargs: 其他传递给plt.plot的参数。

    返回:
        matplotlib.lines.Line2D: 绘制的圆对象。
    """
    angle = np.linspace(0, 2 * np.pi, points)
    x = center[0] + radius * np.cos(angle)
    y = center[1] + radius * np.sin(angle)

    if ax is None:
        ax = plt.gca()

    line = ax.plot(
        x,
        y,
        color=color,
        alpha=alpha,
        linestyle=linestyle,
        linewidth=linewidth,
        label=label,
        marker=marker,
        markersize=markersize,
        markerfacecolor=markerfacecolor,
        markeredgecolor=markeredgecolor,
        markeredgewidth=markeredgewidth,
        **kwargs,
    )

    ax.axis("equal")
    return line[0] if line else None


def draw_circle(
    radius=1,
    color="#0F0",
    alpha=1,
    center=(0, 0),
    points=100,
    linestyle="-",
    linewidth=1,
    label=None,
    marker=None,
    markersize=5,
    markerfacecolor="r",
    markeredgecolor="k",
    markeredgewidth=1,
):
    angle = np.linspace(0, 2 * np.pi, points)
    x = center[0] + radius * np.cos(angle)
    y = center[1] + radius * np.sin(angle)
    plt.axis("equal")
    plt.plot(
        x,
        y,
        color=color,
        alpha=alpha,
        linestyle=linestyle,
        linewidth=linewidth,
        label=label,
        marker=marker,
        markersize=markersize,
        markerfacecolor=markerfacecolor,
        markeredgecolor=markeredgecolor,
        markeredgewidth=markeredgewidth,
    )


# def circle(center, radius, color='b'):
#     angle = np.linspace(0, 2*np.pi, 1000)
#     x = center[0] + radius * np.cos(angle)
#     y = center[1] + radius * np.sin(angle)
#     plt.axis('equal')
#     plt.plot(x, y, color)


def circle_p(center, point, color="b"):
    # 计算圆的半径
    radius = np.sqrt((point[0] - center[0]) ** 2 + (point[1] - center[1]) ** 2)
    # 生成圆上的点
    theta = np.linspace(0, 2 * np.pi, 100)
    x = center[0] + radius * np.cos(theta)
    y = center[1] + radius * np.sin(theta)
    # 绘制图形
    plt.axis("equal")
    plt.plot(x, y, color)


# 椭圆


def ellipse(a=2, b=1, rotation=0, use_degree=True, color="#0f0", alpha=1, center=(0, 0), points=1000):
    """绘制椭圆。

    参数:
        a (float): 半长轴。
        b (float): 半短轴。
        rotation (float): 旋转角度（度或弧度取决于 use_degree），默认为 0。
        use_degree (bool): rotation 是否使用角度制，默认 True（度）。
        color, alpha, center, points: 样式参数。
    """
    theta = np.linspace(0, 2 * np.pi, points)
    angle_rad = np.deg2rad(rotation) if use_degree else rotation
    x = (
        a * np.cos(theta) * np.cos(angle_rad)
        - b * np.sin(theta) * np.sin(angle_rad)
        + center[0]
    )
    y = (
        a * np.cos(theta) * np.sin(angle_rad)
        + b * np.sin(theta) * np.cos(angle_rad)
        + center[1]
    )
    plt.plot(x, y, color, alpha)


# 统一的同心圆函数
def concentric_circles(
    center=(0, 0),
    n=3,
    radius=1,
    param=0.5,
    mode="arithmetic",
    direction="both",
    color="#0F0",
    marker="o",
    alpha=1,
    points=200,
    linestyle="-",
    linewidth=1,
    ax=None,
    show_center=True,
    **kwargs,
):
    """
    绘制同心圆的统一函数（兼容等差与等比序列）。

    参数:
        center (tuple): 圆心坐标。
        n (int): 每一侧的圈数（当 direction=='both' 时每侧各 n 个）。
        radius (float): 基准半径（绘制在中间）。
        param (float): 当 mode=='arithmetic' 时作为公差 d；当 mode=='geometric' 时作为公比 q (>0, q!=0, q!=1 推荐)。
        mode (str): 'arithmetic' 或 'geometric'，选择等差或等比。
        direction (str): 'both'|'out'|'in'，控制向外/向内/双向绘制。
        color, marker, alpha, points, linestyle, linewidth: 绘图样式参数。
        ax (Axes): 可选的 matplotlib 坐标轴对象。
        show_center (bool): 是否绘制圆心标记。
        **kwargs: 其它传入 `ax.plot` 的参数。

    说明:
        - 函数使用与 `circle()` 相同的参数风格（可接收 ax 与样式参数）。
        - 当 direction=='both' 时，会在基准半径两侧各生成 n 个半径值。
    """
    if ax is None:
        ax = plt.gca()

    if show_center:
        ax.plot(center[0], center[1], marker=marker, color=color)

    radii = []
    mode = mode.lower()
    direction = direction.lower()

    if mode.startswith("a"):
        # 等差: param 作为公差 d
        d = float(param)
        if direction == "both":
            for i in range(1, n + 1):
                radii.append(radius + i * d)
                radii.append(radius - i * d)
        elif direction == "out":
            for i in range(1, n + 1):
                radii.append(radius + i * d)
        elif direction == "in":
            for i in range(1, n + 1):
                radii.append(radius - i * d)
    elif mode.startswith("g"):
        # 等比: param 作为公比 q
        q = float(param)
        if q == 0:
            raise ValueError("ratio (param) must be non-zero for geometric mode")
        if direction == "both":
            for i in range(1, n + 1):
                radii.append(radius * (q**i))
                radii.append(radius / (q**i))
        elif direction == "out":
            for i in range(1, n + 1):
                radii.append(radius * (q**i))
        elif direction == "in":
            for i in range(1, n + 1):
                radii.append(radius / (q**i))

    # 确保半径为正且不重复，并保持基准半径也被绘制
    radii = [r for r in radii if r > 0]
    # 加入基准半径并去重、排序（近到远）
    radii.append(radius)
    radii = sorted(set(radii))

    for r in radii:
        angle = np.linspace(0, 2 * np.pi, points)
        x = center[0] + r * np.cos(angle)
        y = center[1] + r * np.sin(angle)
        ax.plot(
            x,
            y,
            color=color,
            alpha=alpha,
            linestyle=linestyle,
            linewidth=linewidth,
            **kwargs,
        )

    ax.axis("equal")


def concentric_ellipse(
    center=(0, 0),
    n=3,
    a=2,
    b=1,
    param=0.5,
    mode="arithmetic",
    direction="both",
    rotation=0,
    color="#0F0",
    marker="o",
    alpha=1,
    points=200,
    linestyle="-",
    linewidth=1,
    ax=None,
    show_center=True,
    **kwargs,
):
    """
    绘制同心椭圆（兼容等差与等比序列）。

    参数:
        center (tuple): 椭圆中心。
        n (int): 每侧的椭圆数量（direction=='both' 时每侧各 n 个）。
        a (float): 基准半长轴。
        b (float): 基准半短轴。
        param (float): 等差时为公差 d；等比时为公比 q。
        mode (str): 'arithmetic' 或 'geometric'。
        direction (str): 'both'|'out'|'in'，控制向外/向内/双向绘制。
        rotation (float): 椭圆旋转角度（度）。
        其它样式参数同 `concentric_circles`。
    """
    if ax is None:
        ax = plt.gca()

    if show_center:
        ax.plot(center[0], center[1], marker=marker, color=color)

    ellipses = []  # 存放 (a_i, b_i)
    mode = mode.lower()
    direction = direction.lower()

    if mode.startswith("a"):
        d = float(param)
        if direction == "both":
            for i in range(1, n + 1):
                ellipses.append((a + i * d, b + i * d))
                ellipses.append((a - i * d, b - i * d))
        elif direction == "out":
            for i in range(1, n + 1):
                ellipses.append((a + i * d, b + i * d))
        elif direction == "in":
            for i in range(1, n + 1):
                ellipses.append((a - i * d, b - i * d))
    elif mode.startswith("g"):
        q = float(param)
        if q == 0:
            raise ValueError("ratio (param) must be non-zero for geometric mode")
        if direction == "both":
            for i in range(1, n + 1):
                ellipses.append((a * (q**i), b * (q**i)))
                ellipses.append((a / (q**i), b / (q**i)))
        elif direction == "out":
            for i in range(1, n + 1):
                ellipses.append((a * (q**i), b * (q**i)))
        elif direction == "in":
            for i in range(1, n + 1):
                ellipses.append((a / (q**i), b / (q**i)))

    # 加入基准椭圆并过滤无效项
    ellipses.append((a, b))
    # 去重（按数值近似）并按面积排序（小到大）
    uniq = {}
    for aa, bb in ellipses:
        if aa > 0 and bb > 0:
            key = (round(float(aa), 12), round(float(bb), 12))
            uniq[key] = (float(aa), float(bb))

    sorted_ellipses = sorted(uniq.values(), key=lambda t: t[0] * t[1])

    theta = np.linspace(0, 2 * np.pi, points)
    angle_rad = np.deg2rad(rotation)

    for aa, bb in sorted_ellipses:
        x = (
            aa * np.cos(theta) * np.cos(angle_rad)
            - bb * np.sin(theta) * np.sin(angle_rad)
            + center[0]
        )
        y = (
            aa * np.cos(theta) * np.sin(angle_rad)
            + bb * np.sin(theta) * np.cos(angle_rad)
            + center[1]
        )
        ax.plot(
            x,
            y,
            color=color,
            alpha=alpha,
            linestyle=linestyle,
            linewidth=linewidth,
            **kwargs,
        )

    ax.axis("equal")

File: cprdspy/CPR_js1/Circles/circle.js
/**
 * CPR 图形库 - JavaScript 版本
 * Circles 模块 - 圆形和椭圆相关函数
 */

/**
 * 绘制圆形
 * @param {Object} options - 配置选项
 * @param {number} options.radius - 半径，默认 1
 * @param {string} options.color - 颜色，默认 "#0F0"
 * @param {number} options.alpha - 透明度，默认 1
 * @param {Array} options.center - 圆心坐标 [x, y]，默认 [0, 0]
 * @param {number} options.points - 点数，默认 100
 * @param {number} options.lineWidth - 线宽，默认 1
 * @param {string} options.label - 标签
 * @returns {Object} 圆形路径数据
 */
function circle({
    radius = 1,
    color = "#0F0",
    alpha = 1,
    center = [0, 0],
    points = 100,
    lineWidth = 1,
    label = null,
} = {}) {
    const pathData = [];
    for (let i = 0; i <= points; i++) {
        const angle = (2 * Math.PI * i) / points;
        const x = center[0] + radius * Math.cos(angle);
        const y = center[1] + radius * Math.sin(angle);
        if (i === 0) {
            pathData.push(`M ${x} ${y}`);
        } else {
            pathData.push(`L ${x} ${y}`);
        }
    }
    pathData.push("Z"); // 闭合路径

    return {
        type: "circle",
        d: pathData.join(" "),
        fill: "none",
        stroke: color,
        "stroke-width": lineWidth,
        "stroke-opacity": alpha,
        label: label,
    };
}

/**
 * 通过圆心和圆上一点绘制圆
 * @param {Array} center - 圆心坐标
 * @param {Array} point - 圆上一点
 * @param {string} color - 颜色
 * @returns {Object} 圆形路径数据
 */
function circleP(center, point, color = "b") {
    const dx = point[0] - center[0];
    const dy = point[1] - center[1];
    const radius = Math.sqrt(dx * dx + dy * dy);
    return circle({ radius, color, center });
}

/**
 * 绘制椭圆
 * @param {Object} options - 配置选项
 * @param {number} options.a - 半长轴，默认 2
 * @param {number} options.b - 半短轴，默认 1
 * @param {number} options.rotation - 旋转角度（度），默认 0
 * @param {string} options.color - 颜色
 * @param {number} options.alpha - 透明度
 * @param {Array} options.center - 中心坐标
 * @param {number} options.points - 点数
 * @param {number} options.lineWidth - 线宽
 * @param {string} options.label - 标签
 * @returns {Object} 椭圆路径数据
 */
function ellipse({
    a = 2,
    b = 1,
    rotation = 0,
    color = "#0f0",
    alpha = 1,
    center = [0, 0],
    points = 1000,
    lineWidth = 1,
    label = null,
} = {}) {
    const angleRad = (rotation * Math.PI) / 180;
    const pathData = [];

    for (let i = 0; i <= points; i++) {
        const theta = (2 * Math.PI * i) / points;
        const x =
            a * Math.cos(theta) * Math.cos(angleRad) -
            b * Math.sin(theta) * Math.sin(angleRad) +
            center[0];
        const y =
            a * Math.cos(theta) * Math.sin(angleRad) +
            b * Math.sin(theta) * Math.cos(angleRad) +
            center[1];
        if (i === 0) {
            pathData.push(`M ${x} ${y}`);
        } else {
            pathData.push(`L ${x} ${y}`);
        }
    }
    pathData.push("Z");

    return {
        type: "ellipse",
        d: pathData.join(" "),
        fill: "none",
        stroke: color,
        "stroke-width": lineWidth,
        "stroke-opacity": alpha,
        label: label,
    };
}

/**
 * 绘制同心圆
 * @param {Object} options - 配置选项
 * @param {Array} options.center - 圆心坐标
 * @param {number} options.n - 每侧圈数
 * @param {number} options.radius - 基准半径
 * @param {number} options.param - 公差（等差）或公比（等比）
 * @param {string} options.mode - "arithmetic" 或 "geometric"
 * @param {string} options.direction - "both", "out", "in"
 * @param {string} options.color - 颜色
 * @param {number} options.alpha - 透明度
 * @param {number} options.points - 点数
 * @param {number} options.lineWidth - 线宽
 * @param {boolean} options.showCenter - 是否显示圆心
 * @returns {Array} 圆形路径数据数组
 */
function concentricCircles({
    center = [0, 0],
    n = 3,
    radius = 1,
    param = 0.5,
    mode = "arithmetic",
    direction = "both",
    color = "#0F0",
    alpha = 1,
    points = 200,
    lineWidth = 1,
    showCenter = true,
} = {}) {
    const paths = [];

    // 添加圆心标记
    if (showCenter) {
        paths.push({
            type: "center",
            cx: center[0],
            cy: center[1],
            fill: color,
            r: 3,
        });
    }

    const radii = [];
    const d = parseFloat(param);
    const q = parseFloat(param);

    if (mode.toLowerCase().startsWith("a")) {
        // 等差数列
        if (direction === "both") {
            for (let i = 1; i <= n; i++) {
                radii.push(radius + i * d);
                radii.push(radius - i * d);
            }
        } else if (direction === "out") {
            for (let i = 1; i <= n; i++) {
                radii.push(radius + i * d);
            }
        } else if (direction === "in") {
            for (let i = 1; i <= n; i++) {
                radii.push(radius - i * d);
            }
        }
    } else if (mode.toLowerCase().startsWith("g")) {
        // 等比数列
        if (q === 0) {
            throw new Error("ratio (param) must be non-zero for geometric mode");
        }
        if (direction === "both") {
            for (let i = 1; i <= n; i++) {
                radii.push(radius * Math.pow(q, i));
                radii.push(radius / Math.pow(q, i));
            }
        } else if (direction === "out") {
            for (let i = 1; i <= n; i++) {
                radii.push(radius * Math.pow(q, i));
            }
        } else if (direction === "in") {
            for (let i = 1; i <= n; i++) {
                radii.push(radius / Math.pow(q, i));
            }
        }
    }

    // 过滤正半径并添加基准半径
    radii.push(radius);
    const uniqueRadii = [...new Set(radii)].filter((r) => r > 0).sort((a, b) => a - b);

    for (const r of uniqueRadii) {
        paths.push(
            circle({
                radius: r,
                color: color,
                alpha: alpha,
                center: center,
                points: points,
                lineWidth: lineWidth,
            })
        );
    }

    return paths;
}

/**
 * 绘制同心椭圆
 * @param {Object} options - 配置选项
 * @param {Array} options.center - 椭圆中心
 * @param {number} options.n - 每侧椭圆数量
 * @param {number} options.a - 基准半长轴
 * @param {number} options.b - 基准半短轴
 * @param {number} options.param - 公差或公比
 * @param {string} options.mode - "arithmetic" 或 "geometric"
 * @param {string} options.direction - "both", "out", "in"
 * @param {number} options.rotation - 旋转角度（度）
 * @param {string} options.color - 颜色
 * @param {number} options.alpha - 透明度
 * @param {number} options.points - 点数
 * @param {number} options.lineWidth - 线宽
 * @param {boolean} options.showCenter - 是否显示中心
 * @returns {Array} 椭圆路径数据数组
 */
function concentricEllipses({
    center = [0, 0],
    n = 3,
    a = 2,
    b = 1,
    param = 0.5,
    mode = "arithmetic",
    direction = "both",
    rotation = 0,
    color = "#0F0",
    alpha = 1,
    points = 200,
    lineWidth = 1,
    showCenter = true,
} = {}) {
    const paths = [];

    if (showCenter) {
        paths.push({
            type: "center",
            cx: center[0],
            cy: center[1],
            fill: color,
            r: 3,
        });
    }

    const ellipses = [];
    const d = parseFloat(param);
    const q = parseFloat(param);

    if (mode.toLowerCase().startsWith("a")) {
        if (direction === "both") {
            for (let i = 1; i <= n; i++) {
                ellipses.push([a + i * d, b + i * d]);
                ellipses.push([a - i * d, b - i * d]);
            }
        } else if (direction === "out") {
            for (let i = 1; i <= n; i++) {
                ellipses.push([a + i * d, b + i * d]);
            }
        } else if (direction === "in") {
            for (let i = 1; i <= n; i++) {
                ellipses.push([a - i * d, b - i * d]);
            }
        }
    } else if (mode.toLowerCase().startsWith("g")) {
        if (q === 0) {
            throw new Error("ratio (param) must be non-zero for geometric mode");
        }
        if (direction === "both") {
            for (let i = 1; i <= n; i++) {
                ellipses.push([a * Math.pow(q, i), b * Math.pow(q, i)]);
                ellipses.push([a / Math.pow(q, i), b / Math.pow(q, i)]);
            }
        } else if (direction === "out") {
            for (let i = 1; i <= n; i++) {
                ellipses.push([a * Math.pow(q, i), b * Math.pow(q, i)]);
            }
        } else if (direction === "in") {
            for (let i = 1; i <= n; i++) {
                ellipses.push([a / Math.pow(q, i), b / Math.pow(q, i)]);
            }
        }
    }

    ellipses.push([a, b]);

    // 去重并排序
    const uniqueMap = new Map();
    for (const [aa, bb] of ellipses) {
        if (aa > 0 && bb > 0) {
            const key = `${aa.toFixed(12)},${bb.toFixed(12)}`;
            uniqueMap.set(key, [aa, bb]);
        }
    }

    const sortedEllipses = [...uniqueMap.values()].sort(
        ([a1, b1], [a2, b2]) => a1 * b1 - a2 * b2
    );

    for (const [aa, bb] of sortedEllipses) {
        paths.push(
            ellipse({
                a: aa,
                b: bb,
                rotation: rotation,
                color: color,
                alpha: alpha,
                center: center,
                points: points,
                lineWidth: lineWidth,
            })
        );
    }

    return paths;
}

// 导出模块
export { circle, circleP, ellipse, concentricCircles, concentricEllipses };
